### PR TITLE
support libcurl

### PR DESCRIPTION
**Pull request by tongheihei**

### CODE REVIEW

File: cmake/FindLibcurl.cmake [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#     find_package(Libcurl)
+#
+# Variables used by this module, they can change the default behaviour and need
+# to be set before calling find_package:
+#
+# Variables defined by this module:
+#
+#  Libevent_FOUND            System has Libevent, include and lib dirs found
+#  Libevent_INCLUDE_DIR      The Libevent includes directories.
+#  Libevent_LIBRARY          The Libevent library.
```
Comment (darionyaphet) [inline cmake/FindLibcurl.cmake:14]: Libcurl ?You need to fix it

File: src/common/http/HttpClient.cpp [hunk 2]
```diff
@@ -1,36 +1,113 @@
-/* Copyright (c) 2019 vesoft inc. All rights reserved.
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:1]: No need update

File: src/common/http/HttpClient.cpp [hunk 3]
```diff
@@ -61,12 +138,51 @@ StatusOr<std::string> HttpClient::put(const std::string& path, const std::string
...
-    if (result.ok()) {
-        return result.value();
-    } else {
-        return Status::Error(folly::stringPrintf("Http Put Failed: %s", path.c_str()));
+    // auto result = nebula::ProcessUtils::runCommand(command.c_str());
+    // if (result.ok()) {
+    //     return result.value();
+    // } else {
+    //     return Status::Error(folly::stringPrintf("Http Put Failed: %s", path.c_str()));
+    // }
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:146]: remove

**Comment by bright-starry-sky** [comment]
What's the purpose?

**Comment by darionyaphet** [comment]
> What's the purpose?

Replace execution command line

File: cmake/FindLibcurl.cmake [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# Variables defined by this module:
+#
+#  Libcurl_FOUND            System has Libcurl, include and lib dirs found
+#  Libcurl_INCLUDE_DIR      The Libcurl includes directories.
+#  Libcurl_LIBRARY          The Libcurl library.
+
+find_path(Libcurl_INCLUDE_DIR NAMES curl/curl.h PATHS /usr/local/include)
+find_library(Libcurl_LIBRARY NAMES  libcurl curl PATHS /usr/local/lib)
+message(STATUS "********Libcurl_INCLUDE_DIR is ${Libcurl_INCLUDE_DIR}")
+message(STATUS "********Libcurl_LIBRARY is ${Libcurl_LIBRARY}")
```
Comment (darionyaphet) [inline cmake/FindLibcurl.cmake:19]: please remove it

Comment (tongheihei) [reply cmake/FindLibcurl.cmake]: I've fixed all of these issues you have mentioned!Thank you !

File: src/common/http/HttpClient.cpp [hunk 5]
```diff
@@ -76,25 +170,52 @@ StatusOr<std::string> HttpClient::put(const std::string& path, const folly::dyna
...
+              curl_easy_strerror(res));
+        }
+        StatusOr<std::string> result = out.str();
+        if (result.ok()) {
+            return result.value();
+        } else {
+            return Status::Error(folly::stringPrintf("Http Get Failed: %s", path.c_str()));
+        }
     }
+    std::cout << "libcurl failed" << std::endl;
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:217]: ?

File: cmake/nebula/ThirdPartyConfig.cmake [hunk 6]
```diff
@@ -1,4 +1,4 @@
-message(">>>> Configuring third party for '${PROJECT_NAME}' <<<<")
+message(STATUS">>>> Configuring third party for '${PROJECT_NAME}' <<<<")
```
Comment (darionyaphet) [inline cmake/nebula/ThirdPartyConfig.cmake:1]: ?

File: cmake/nebula/ThirdPartyConfig.cmake [hunk 7]
```diff
@@ -16,6 +17,7 @@ if(${DISABLE_CXX11_ABI})
     endif()
 else()
     if("${NEBULA_THIRDPARTY_ROOT}" STREQUAL "")
+        message(STATUS"!!!!!!!!!!!!!!!!!!!!1111111")
```
Comment (darionyaphet) [inline cmake/nebula/ThirdPartyConfig.cmake:20]: ?

File: cmake/nebula/ThirdPartyConfig.cmake [hunk 8]
```diff
@@ -36,10 +38,19 @@ if(NOT ${NEBULA_THIRDPARTY_ROOT} STREQUAL "")
...
+    List(INSERT CMAKE_INCLUDE_PATH 0 /usr/local/include)
+    List(INSERT CMAKE_LIBRARY_PATH 0 /usr/local/lib)
+    include_directories(SYSTEM ${NEBULA_THIRDPARTY_ROOT}/include /usr/local/include)
+    message(STATUS "Build info of nebula third party:\n${third_party_build_info}")
+    #change of tongheihei
+    #include_directories(/usr/local/include)
     link_directories(
         ${NEBULA_THIRDPARTY_ROOT}/lib
         ${NEBULA_THIRDPARTY_ROOT}/lib64
+        #change of tongheiehei
```
Comment (darionyaphet) [inline cmake/nebula/ThirdPartyConfig.cmake:51]: ?

File: cmake/nebula/ThirdPartyConfig.cmake [hunk 9]
```diff
@@ -36,10 +38,19 @@ if(NOT ${NEBULA_THIRDPARTY_ROOT} STREQUAL "")
...
     list(INSERT CMAKE_LIBRARY_PATH 0 ${NEBULA_THIRDPARTY_ROOT}/lib64)
     list(INSERT CMAKE_PROGRAM_PATH 0 ${NEBULA_THIRDPARTY_ROOT}/bin)
-    include_directories(SYSTEM ${NEBULA_THIRDPARTY_ROOT}/include)
+    #change of tongheihei
+    List(INSERT CMAKE_INCLUDE_PATH 0 /usr/local/include)
+    List(INSERT CMAKE_LIBRARY_PATH 0 /usr/local/lib)
+    include_directories(SYSTEM ${NEBULA_THIRDPARTY_ROOT}/include /usr/local/include)
+    message(STATUS "Build info of nebula third party:\n${third_party_build_info}")
+    #change of tongheihei
+    #include_directories(/usr/local/include)
```
Comment (darionyaphet) [inline cmake/nebula/ThirdPartyConfig.cmake:47]: ?

File: cmake/nebula/ThirdPartyConfig.cmake [hunk 10]
```diff
@@ -149,5 +161,5 @@ if (NOT ENABLE_JEMALLOC OR ENABLE_ASAN OR ENABLE_UBSAN)
 else()
     set(JEMALLOC_LIB jemalloc)
 endif()
-
-message(">>>> Configuring third party for '${PROJECT_NAME}' done <<<<")
+message(STATUS"****************************************"${CMAKE_BINARY_DIR})
```
Comment (darionyaphet) [inline cmake/nebula/ThirdPartyConfig.cmake:164]: ?

File: src/common/http/HttpClient.cpp [hunk 11]
```diff
@@ -3,34 +3,98 @@
...
+        } while (len == 1024);
+        fclose(f);
+        StatusOr<std::string> result = out.str();
+        curl_easy_cleanup(curl);
+        if (result.ok()) {
+            return result.value();
+        } else {
+            return Status::Error(folly::stringPrintf("Http Get Failed: %s", path.c_str()));
+        }
     }
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:54]: Suitable for C++

File: src/common/http/HttpClient.cpp [hunk 12]
```diff
@@ -76,25 +170,52 @@ StatusOr<std::string> HttpClient::put(const std::string& path, const folly::dyna
...
+        fclose(f);
+        curl_slist_free_all(headers);
+        curl_easy_cleanup(curl);
+        curl_global_cleanup();
+        if (res != CURLE_OK) {
+            fprintf(stderr, "curl_easy_perform() failed: %s\n",
+              curl_easy_strerror(res));
+        }
+        StatusOr<std::string> result = out.str();
+        if (result.ok()) {
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:211]: Suitable for C++

File: src/common/http/HttpClient.cpp [hunk 13]
```diff
@@ -3,40 +3,99 @@
...
-        return result.value();
-    } else {
-        return Status::Error(folly::stringPrintf("Http Post Failed: %s", path.c_str()));
+    CURL *curl;
+    CURLcode res;
+    struct curl_slist *my_curl_list = nullptr;
+    int len = header.length();
+    int now_total = 0;
+    std::string newString = "";
+   for ( int i = 0 ; i < len ; i++ ) {
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:66]: alignment and space

Comment (tongheihei) [reply src/common/http/HttpClient.cpp]: I've fixed it,thank you !

File: src/common/http/HttpClient.cpp [hunk 14]
```diff
@@ -50,23 +110,47 @@ StatusOr<std::string> HttpClient::post(const std::string& path, const folly::dyn
...
+            }
+            newString+=header[i];
     }
+    curl = curl_easy_init();
+    if (curl) {
+    curl_easy_setopt(curl, CURLOPT_URL, path.c_str());
+    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, my_curl_list);
+    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "put");
+    res = curl_easy_perform(curl);
+    curl_slist_free_all(my_curl_list);
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:146]: indent ?

File: src/common/http/HttpClient.cpp [hunk 15]
```diff
@@ -76,25 +160,41 @@ StatusOr<std::string> HttpClient::put(const std::string& path, const folly::dyna
...
+        curl_easy_cleanup(curl);
+        curl_global_cleanup();
+        if (result.ok()) {
+            return result.value();
+        } else {
+            return Status::Error(folly::stringPrintf("Http Get Failed: %s", path.c_str()));
+        }
     }
+    free(chunk.memory);
+    return "libcurl failed";
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:197]: If  the function failed it should return an status

File: src/common/http/HttpClient.cpp (hunk 14 above)
Comment (tongheihei) [reply src/common/http/HttpClient.cpp]: I've reviewed all the intents of the code now.

File: src/common/http/HttpClient.cpp (hunk 15 above)
Comment (tongheihei) [reply src/common/http/HttpClient.cpp]: I've changed it, thank you.

File: src/common/http/HttpClient.cpp [hunk 16]
```diff
@@ -3,40 +3,100 @@
...
+        curl_easy_setopt(curl, CURLOPT_URL, path.c_str());
+        curl_easy_setopt(curl, CURLOPT_POST, 1L);
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, my_curl_list);
+        res = curl_easy_perform(curl);
+        curl_slist_free_all(my_curl_list);
+        if (res != CURLE_OK) {
+            fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+            return Status::Error(folly::stringPrintf("Http Get Failed: %s", path.c_str()));
+        }
+        curl_easy_cleanup(curl);
```
Comment (critical27) [inline src/common/http/HttpClient.cpp:92]: Always return Status::Error?

File: src/common/http/HttpClient.cpp [hunk 17]
```diff
@@ -50,23 +110,47 @@ StatusOr<std::string> HttpClient::post(const std::string& path, const folly::dyn
...
-        return result.value();
-    } else {
-        return Status::Error(folly::stringPrintf("Http Put Failed: %s", path.c_str()));
+    CURL *curl;
+    CURLcode res;
+    struct curl_slist *my_curl_list = nullptr;
+    int len = header.length();
+    int now_total = 0;
+    std::string newString = "";
+    for ( int i = 0 ; i < len ; i++ ) {
```
Comment (critical27) [inline src/common/http/HttpClient.cpp:126]: Would you explain what are we doing here?

File: src/common/http/HttpClient.cpp [hunk 18]
```diff
@@ -3,40 +3,100 @@
...
+    mem->size += realsize;
+    mem->memory[mem->size] = 0;
+    return realsize;
+}
+ StatusOr<std::string> HttpClient::get(const std::string& path, const std::string& /*options*/) {
+    CURL *curl;
+    CURLcode res;
+    curl = curl_easy_init();
+    struct MemoryStruct chunk;
+    chunk.memory = reinterpret_cast<char *> (malloc(1));
```
Comment (critical27) [inline src/common/http/HttpClient.cpp:36]: Maybe we should not use any `malloc/free/realloc`, which is error prone.

Better take a look at other examples, e.g. https://curl.se/libcurl/c/htmltitle.html and https://gist.github.com/alghanmi/c5d7b761b2c9ab199157.

**Review by critical27 — COMMENTED** [review]
It seems the code is way more complicated than before 😂

File: src/common/http/HttpClient.cpp (hunk 16 above)
Comment (tongheihei) [reply src/common/http/HttpClient.cpp:92]: Only when the curl_ok code not eqauls to zero which means there are some issues in this libcurl task,and for that the result string will be NULL so I return Status::Error.The code also returns result.value() when the libcurl task is correct.

File: src/common/http/HttpClient.cpp (hunk 17 above)
Comment (tongheihei) [reply src/common/http/HttpClient.cpp:126]: The http-request may have header data,And it's always a string which struct as JSON. The part of codes is use to get the header of the JSON string and put it into Libcurl's  CURLOPT_HTTPHEADER.

File: src/common/http/HttpClient.cpp [hunk 19]
```diff
@@ -3,40 +3,84 @@
...
-        return Status::Error(folly::stringPrintf("Http Get Failed: %s", path.c_str()));
+static size_t WriteCallback(void *contents, size_t size, size_t nmemb, void *userp) {
+    reinterpret_cast<std::string*>(userp)->append(reinterpret_cast<char*>(contents), size * nmemb);
+    return size * nmemb;
+}
+ StatusOr<std::string> HttpClient::get(const std::string& path, const std::string& /*options*/) {
+    CURL *curl;
+    CURLcode res;
+    curl = curl_easy_init();
+    std::string return_string;
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:25]: CamelCase

File: src/common/http/HttpClient.cpp [hunk 20]
```diff
@@ -3,40 +3,84 @@
...
-        folly::stringPrintf("/usr/bin/curl -X POST %s \"%s\"", header.c_str(), path.c_str());
-    LOG(INFO) << "HTTP POST Command: " << command;
-    auto result = nebula::ProcessUtils::runCommand(command.c_str());
-    if (result.ok()) {
-        return result.value();
-    } else {
-        return Status::Error(folly::stringPrintf("Http Post Failed: %s", path.c_str()));
+    CURL *curl;
+    CURLcode res;
+    struct curl_slist *my_curl_list = nullptr;
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:47]: ditto

File: src/common/http/HttpClient.cpp [hunk 21]
```diff
@@ -3,40 +3,84 @@
...
+    }
+    curl = curl_easy_init();
+    if (curl) {
+        curl_easy_setopt(curl, CURLOPT_URL, path.c_str());
+        curl_easy_setopt(curl, CURLOPT_POST, 1L);
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, my_curl_list);
+        res = curl_easy_perform(curl);
+        curl_slist_free_all(my_curl_list);
+        if (res != CURLE_OK) {
+            fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:73]: LOG(ERROR)

File: src/common/http/HttpClient.cpp [hunk 22]
```diff
@@ -3,40 +3,84 @@
...
+            if (now_total%2 == 1) {
+                newString = "";
+                continue;
+            }
+            if (now_total%2 == 0) {
+                my_curl_list = curl_slist_append(my_curl_list, newString.c_str());
+            }
+            continue;
+        }
+            newString+=header[i];
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:63]: need space

File: src/common/http/HttpClient.cpp [hunk 23]
```diff
@@ -3,40 +3,90 @@
...
-    auto result = nebula::ProcessUtils::runCommand(command.c_str());
-    if (result.ok()) {
-        return result.value();
-    } else {
-        return Status::Error(folly::stringPrintf("Http Post Failed: %s", path.c_str()));
+    CURL *curl;
+    CURLcode res;
+    struct curl_slist *headers = nullptr;
+    auto len = header.length();
+    auto nowTotal = 0;
```
Comment (darionyaphet) [inline src/common/http/HttpClient.cpp:49]: int32_t

**Comment by bright-starry-sky** [comment]
HttpClient::delete ？

File: src/common/http/HttpClient.cpp [hunk 24]
```diff
@@ -3,40 +3,90 @@
...
+        curl_easy_setopt(curl, CURLOPT_URL, path.c_str());
+        curl_easy_setopt(curl, CURLOPT_POST, 1L);
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &returnString);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+        res = curl_easy_perform(curl);
+        curl_slist_free_all(headers);
+        StatusOr<std::string> result = returnString;
+        curl_easy_cleanup(curl);
+        LOG(INFO) << "HTTP return Code: " << res;
```
Comment (bright-starry-sky) [inline src/common/http/HttpClient.cpp:77]: VLOG ？

File: src/common/http/HttpClient.cpp [hunk 25]
```diff
@@ -3,40 +3,90 @@
...
+        if (result.ok()) {
+            return result.value();
+        } else {
+            return Status::Error(folly::stringPrintf("Http Post Failed: %s", path.c_str()));
+        }
```
Comment (bright-starry-sky) [inline src/common/http/HttpClient.cpp:82]: return returnString

File: src/common/http/HttpClient.cpp [hunk 26]
```diff
@@ -3,40 +3,90 @@
...
+        newString += header[i];
+    }
+    curl = curl_easy_init();
+    if (curl) {
+        curl_easy_setopt(curl, CURLOPT_URL, path.c_str());
+        curl_easy_setopt(curl, CURLOPT_POST, 1L);
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &returnString);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+        res = curl_easy_perform(curl);
```
Comment (bright-starry-sky) [inline src/common/http/HttpClient.cpp:73]: check ret != CURLE_OK

**Review by tongheihei — COMMENTED** [review]
> What's the purpose?

Using an api libcurl http transfer instead of curl.

File: cmake/FindLibcurl.cmake [hunk 27]
```diff
@@ -15,7 +15,8 @@
 
 find_path(Libcurl_INCLUDE_DIR NAMES curl/curl.h )
 find_library(Libcurl_LIBRARY NAMES  libcurl curl )
```
Comment (darionyaphet) [inline cmake/FindLibcurl.cmake:17]: extra space

**Pull request closed without merging**